### PR TITLE
Fix issue with Network interfaces and an instance-level security groups

### LINKED ISSUE #1188: Network interfaces and an instance-level security groups may not be specified on the same request
It seems the last commit broke assigning security groups to an instance:

```
resource "aws_instance" "nat" {
  ami = "${var.nat_ami}"
  instance_type = "m1.small"
  security_groups = ["${aws_security_group.nat.id}"]
  subnet_id = "${aws_subnet.dmz_subnet.id}"
  associate_public_ip_address = true
  source_dest_check = false
  tags {
    Name = "${concat(var.vpc_name,"-nat")}"
  }
}
```

Output:

```
Error applying plan:

1 error(s) occurred:

* 1 error(s) occurred:

* 1 error(s) occurred:

* 1 error(s) occurred:

* Error launching source instance: Network interfaces and an instance-level security groups may not be specified on the same request

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
```

If I hash out the security groups parameter it works fine.

### PR DESCRIPTION
**Pull request by catsby**
This should fix #1188, pending current acceptance tests and a new one to account for the issue there.

The configuration I tested:

```
provider "aws" {
  region="us-west-2"
}

resource "aws_vpc" "foo" {
  cidr_block = "10.1.0.0/16"
}

resource "aws_security_group" "tf_test_foo" {
  name = "tf_test_foo"
  description = "foo"
  vpc_id="${aws_vpc.foo.id}"

  ingress {
    protocol = "icmp"
    from_port = -1
    to_port = -1
    cidr_blocks = ["0.0.0.0/0"]
  }
}

resource "aws_subnet" "foo" {
  cidr_block = "10.1.1.0/24"
  vpc_id = "${aws_vpc.foo.id}"
}

resource "aws_instance" "nat" {
  ami = "ami-21f78e11"
  instance_type = "t1.micro"
  security_groups = ["${aws_security_group.tf_test_foo.id}"]
  subnet_id = "${aws_subnet.foo.id}"
  associate_public_ip_address = true
  source_dest_check = false
  tags {
    Name = "testing-vpc-subnet-things"
  }
}
```

### CODE REVIEW

**Comment by catsby** [comment]
cc @phinze

**Comment by jaxxstorm** [comment]
I think you need to associate an elasticip as well:

```
resource "aws_eip" "nat" {
  instance = "${aws_instance.nat.id}"
  vpc = true
}
```

That was part of my config. It might not be relevant to the bug, but I know some of the stuff with API issue I read mentioned associating elasticips in there. Definitely worth sanity checking I think.

**Comment by mitchellh** [comment]
LGTM. It is good we're finding these instance issues, thanks @jaxxstorm :)

**Comment by catsby** [comment]
@jaxxstorm k, trying that. Do you also have an internet gateway? I'm having issues there...

**Comment by jaxxstorm** [comment]
yep, I sure do:

```
resource "aws_internet_gateway" "gw" {
  vpc_id = "${var.vpc_id}"

  tags {
    Name = "${concat(var.vpc_name,"-gateway")}"
  }
}
```

**Comment by kendawg2** [comment]
@jaxxstorm I have this and it works:

resource "aws_internet_gateway" "gw" {
    vpc_id = "${aws_vpc.idc.id}"

    tags {
        Name = "${var.environment}-Internet Gateway"
    }
}

Not really any different except I am not using concat() which really isn't needed for that anymore.

**Comment by catsby** [comment]
Updated the config example:

```
provider "aws" {
  region="us-west-2"
}

resource "aws_internet_gateway" "gw" {
    vpc_id = "${aws_vpc.foo.id}"
    tags {
        Name = "main"
    }
}

resource "aws_vpc" "foo" {
  cidr_block = "10.1.0.0/16"
}

resource "aws_security_group" "tf_test_foo" {
  name = "tf_test_foo"
  description = "foo"
  vpc_id="${aws_vpc.foo.id}"

  ingress {
    protocol = "icmp"
    from_port = -1
    to_port = -1
    cidr_blocks = ["0.0.0.0/0"]
  }
}

resource "aws_subnet" "foo" {
  cidr_block = "10.1.1.0/24"
  vpc_id = "${aws_vpc.foo.id}"
}

resource "aws_instance" "nat" {
  ami = "ami-21f78e11"
  instance_type = "t1.micro"
  security_groups = ["${aws_security_group.tf_test_foo.id}"]
  subnet_id = "${aws_subnet.foo.id}"
  associate_public_ip_address = true
  source_dest_check = false
  tags {
    Name = "testing-vpc-subnet-things"
  }
}

resource "aws_eip" "nat" {
  instance = "${aws_instance.nat.id}"
  vpc = true
}
```

Things seem to work as expected. I'm going to make an acceptance test then get this merged!

**Comment by phinze** [comment]
LGTM

**Comment by phinze** [comment]
(Actually we're debugging an issue with the destroys in the test... stand by...)

terraform_test_graph

File: builtin/providers/aws/resource_aws_instance_test.go [hunk 1]
```diff
@@ -533,3 +552,48 @@ resource "aws_instance" "foo" {
 	private_ip = "10.1.1.42"
 }
 `
+
+const testAccInstanceNetworkInstanceSecurityGroups = `
+resource "aws_internet_gateway" "gw" {
+  vpc_id = "${aws_vpc.foo.id}"
+	depends_on = ["aws_eip.foo_eip"]
```
Comment (phinze) [inline builtin/providers/aws/resource_aws_instance_test.go]: Ok, two things: (1) remove this dependency

File: builtin/providers/aws/resource_aws_instance_test.go [hunk 2]
```diff
@@ -533,3 +552,48 @@ resource "aws_instance" "foo" {
...
+  cidr_block = "10.1.1.0/24"
+  vpc_id = "${aws_vpc.foo.id}"
+}
+
+resource "aws_instance" "foo_instance" {
+  ami = "ami-21f78e11"
+  instance_type = "t1.micro"
+  security_groups = ["${aws_security_group.tf_test_foo.id}"]
+  subnet_id = "${aws_subnet.foo.id}"
+  associate_public_ip_address = true
```
Comment (phinze) [inline builtin/providers/aws/resource_aws_instance_test.go]: Alright actually three things: (2) add IGW dependency here:

```
depends_on = ["aws_internet_gateway.gw"]
```

File: builtin/providers/aws/resource_aws_instance_test.go [hunk 3]
```diff
@@ -533,3 +552,48 @@ resource "aws_instance" "foo" {
...
+}
+
+resource "aws_eip" "foo_eip" {
+  instance = "${aws_instance.foo_instance.id}"
+  vpc = true
```
Comment (phinze) [inline builtin/providers/aws/resource_aws_instance_test.go]: (3) And here:

```
depends_on = ["aws_internet_gateway.gw"]
```

**Comment by phinze** [comment]
Logic for new dependency suggestions:

 * Internet Gateway must be present for instance with public IP address to launch in a VPC, and that instance must be gone by the time the IGW is deleted
 * Internet Gateway must be present for EIP to be attached in a VPC, and that EIP must be detached before IGW is deleted

**Comment by catsby** [comment]
@phinze I think we're good here now, yeah?

**Comment by phinze** [comment]
@catsby Yep this is good to go.

**Pull request merged**